src/storage/storages.rs

```diff
@@ -194,3 +197,138 @@ impl<T: Default> UnprotectedStorage<T> for NullStorage<T> {
         Default::default()
     }
 }
+
+/// Wrapper storage that stores modifications to components in a bitset.
+///
+/// **Note: Never use `.iter()` on a mutable component storage that uses this.**
+///
+///# Example Usage:
+/// 
+/// ```rust
+/// extern crate specs;
+/// use specs::prelude::*;
+/// 
+/// pub struct Comp(u32);
+/// impl Component for Comp {
+///     // `TrackedStorage` acts as a wrapper around another storage.
+///     // You can put any store inside of here (e.g. HashMapStorage, VecStorage, etc.)
+///     type Storage = TrackedStorage<Comp, VecStorage<Comp>>;
+/// }
+/// 
+/// pub struct CompSystem;
+/// impl<'a> System<'a> for CompSystem {
+///     type SystemData = WriteStorage<'a, Comp>;
+///     fn run(&mut self, mut comps: WriteStorage<'a, Comp>) {
+///         // Iterates over all components like normal.
+///         for comp in (&comps).join() {
+///             // ...
+///         }
+/// 
+///         // **Never do this**
+///         // This will flag all components as modified regardless of whether the inner loop
+///         // did modify their data.
+///         for comp in (&mut comps).join() {
+///             // ...
+///         }
+/// 
+///         // Instead do something like:
+///         for mut entry in (&comps.check()).join() {
+///             if true { // check whether this component should be modified.
+///                 let mut comp = comps.get_mut_unchecked(&mut entry);
+///                 // ...
+///             }
+///         }
+/// 
+///         // To iterate over the flagged/modified components:
+///         for flagged_comp in ((&comps).open().1).join() {
+///             // ...
+///         }
+///
+///         // Clears the tracked storage every frame with this system.
+///         (&mut comps).open().1.clear_flags();
+///     }
+/// }
+///# fn main() { }
+/// ```
+pub struct TrackedStorage<C, T> {
```

Comment:
I just had an idea:

Given that both this storage and `ChangeStorage` track the mutation of their components, maybe we should rename this one to `FlaggedStorage`?